browser/brave_rewards/creator_detection_script_injector.cc

```diff
@@ -0,0 +1,182 @@
+    {{"github.com", IDR_CREATOR_DETECTION_GITHUB_BUNDLE_JS},
+     {"www.github.com", IDR_CREATOR_DETECTION_GITHUB_BUNDLE_JS},
+     {"gist.github.com", IDR_CREATOR_DETECTION_GITHUB_BUNDLE_JS},
+     {"reddit.com", IDR_CREATOR_DETECTION_REDDIT_BUNDLE_JS},
+     {"www.reddit.com", IDR_CREATOR_DETECTION_REDDIT_BUNDLE_JS},
+     {"twitch.tv", IDR_CREATOR_DETECTION_TWITCH_BUNDLE_JS},
+     {"www.twitch.tv", IDR_CREATOR_DETECTION_TWITCH_BUNDLE_JS},
+     {"twitter.com", IDR_CREATOR_DETECTION_TWITTER_BUNDLE_JS},
+     {"x.com", IDR_CREATOR_DETECTION_TWITTER_BUNDLE_JS},
+     {"vimeo.com", IDR_CREATOR_DETECTION_VIMEO_BUNDLE_JS},
+     {"www.youtube.com", IDR_CREATOR_DETECTION_YOUTUBE_BUNDLE_JS},
+     {"m.youtube.com", IDR_CREATOR_DETECTION_YOUTUBE_BUNDLE_JS}});
+
+std::string LoadScriptResource(int id) {
+  auto& bundle = ui::ResourceBundle::GetSharedInstance();
+  if (bundle.IsGzipped(id)) {
+    return bundle.LoadDataResourceString(id);
+  }
+  return std::string(bundle.GetRawDataResource(id));
+}
+
+std::optional<std::string> GetDetectionScript(content::RenderFrameHost* rfh) {
+  // Only run scripts for the main frame.
+  if (!rfh || !rfh->IsInPrimaryMainFrame()) {
+    return std::nullopt;
+  }
+
+  // Only run scripts if the creator detection feature is enabled.
+  if (!base::FeatureList::IsEnabled(
+          features::kPlatformCreatorDetectionFeature)) {
+    return std::nullopt;
+  }
+
+  auto* profile = Profile::FromBrowserContext(rfh->GetBrowserContext());
+
+  // Only run scripts if the Rewards service is available for this profile.
+  if (!IsSupportedForProfile(profile)) {
+    return std::nullopt;
+  }
+
+  // Only run scripts if the user has enabled Brave Rewards.
+  if (!profile || !profile->GetPrefs()->GetBoolean(prefs::kEnabled)) {
+    return std::nullopt;
+  }
+
+  // Only run scripts for known "media platform" sites.
+  GURL url = rfh->GetLastCommittedURL();
+  if (!IsMediaPlatformURL(url)) {
+    return std::nullopt;
+  }
+
+  // Only run scripts when there is an exact hostname match.
+  auto iter = kScriptMap.find(url.host_piece());
+  if (iter == kScriptMap.end()) {
+    return std::nullopt;
+  }
+
+  return LoadScriptResource(iter->second);
+}
+
+}  // namespace
+
+CreatorDetectionScriptInjector::CreatorDetectionScriptInjector() = default;
+CreatorDetectionScriptInjector::~CreatorDetectionScriptInjector() = default;
+
+void CreatorDetectionScriptInjector::MaybeInjectScript(
+    content::RenderFrameHost* rfh) {
+  injector_.reset();
+  injector_host_token_ = content::GlobalRenderFrameHostToken();
+
+  if (!rfh) {
+    return;
+  }
+
+  auto script_source = GetDetectionScript(rfh);
+  if (!script_source) {
+    return;
+  }
+
+  injector_host_token_ = rfh->GetGlobalFrameToken();
+  rfh->GetRemoteAssociatedInterfaces()->GetInterface(&injector_);
+
+  bool verbose_logging =
+      base::FeatureList::IsEnabled(features::kVerboseLoggingFeature);
+
+  // Add a `braveRewards` global variable to the isolated world.
```

Comment:
I'm a bit confused about why we're injecting the script separately from running the function. Why don't we just wrap the script with a call to the function and return the result? This seems to complicate the process and it's not clear why we need to do this in two steps?

Reply:
Sorry I missed this on the first pass

Reply:
The bundled scripts aren't small, and it would be more efficient to inject the bundle once and then call it as needed (e.g. for subsequent same-doc navigations). I'm not sure if it would be an overall improvement to execute the bundle each time.